File: src/apps/daemon/internal/networking/dns/tunnel_probe_evaluator.go
package dns

import (
	"math"
	"sort"
	"strings"
)

// AttemptMetric captures latency and result of a single probe attempt.
type AttemptMetric struct {
	Success       bool   `json:"success"`
	DurationMs    int64  `json:"durationMs"`
	ErrorCategory string `json:"errorCategory,omitempty"`
}

// TCPProbeResult encapsulates TCP connectivity and consistency.
type TCPProbeResult struct {
	Success       bool            `json:"success"`
	Attempts      []AttemptMetric `json:"attempts"`
	MedianRTTMs   int64           `json:"medianRttMs"`
	Consistency   float64         `json:"consistency"`
	ErrorCategory string          `json:"errorCategory,omitempty"`
}

// TLSProbeResult stores TLS negotiation outcomes.
type TLSProbeResult struct {
	Success         bool     `json:"success"`
	HandshakeMs     int64    `json:"handshakeMs"`
	Version         string   `json:"version,omitempty"`
	CipherSuite     string   `json:"cipherSuite,omitempty"`
	ALPN            string   `json:"alpn,omitempty"`
	CertificateCN   string   `json:"certificateCn,omitempty"`
	CertificateSANs []string `json:"certificateSans,omitempty"`
	Verified        bool     `json:"verified"`
	ErrorCategory   string   `json:"errorCategory,omitempty"`
}

// HTTPProbeItem records individual HTTP probe requests.
type HTTPProbeItem struct {
	Method        string `json:"method"`
	URL           string `json:"url"`
	StatusCode    int    `json:"statusCode"`
	DurationMs    int64  `json:"durationMs"`
	Redirected    bool   `json:"redirected"`
	ErrorCategory string `json:"errorCategory,omitempty"`
}

// HTTPProbeResult records aggregated HTTP probing outcomes.
type HTTPProbeResult struct {
	Success bool            `json:"success"`
	Probes  []HTTPProbeItem `json:"probes"`
}

// WSProbeResult captures WebSocket handshake attempts.
type WSProbeResult struct {
	Success       bool   `json:"success"`
	StatusCode    int    `json:"statusCode"`
	DurationMs    int64  `json:"durationMs"`
	ErrorCategory string `json:"errorCategory,omitempty"`
}

// UDPProbeResult holds UDP reachability observations.
type UDPProbeResult struct {
	Reachable     bool            `json:"reachable"`
	Attempts      []AttemptMetric `json:"attempts"`
	ErrorCategory string          `json:"errorCategory,omitempty"`
}

// QUICProbeResult tracks QUIC/HTTP3 handshake status.
type QUICProbeResult struct {
	Success       bool   `json:"success"`
	HandshakeMs   int64  `json:"handshakeMs"`
	ALPN          string `json:"alpn,omitempty"`
	ErrorCategory string `json:"errorCategory,omitempty"`
}

// DNSProbeResult tracks resolver UDP/TCP answers.
type DNSProbeResult struct {
	UDPResponsive bool            `json:"udpResponsive"`
	TCPResponsive bool            `json:"tcpResponsive"`
	Answers       []string        `json:"answers,omitempty"`
	Attempts      []AttemptMetric `json:"attempts"`
	ErrorCategory string          `json:"errorCategory,omitempty"`
}

// ProbeMetrics summarizes network path latency, variance, packet loss, and stability.
type ProbeMetrics struct {
	RTTMs              int64   `json:"rttMs"`
	JitterMs           int64   `json:"jitterMs"`
	PacketLossEstimate float64 `json:"packetLossEstimate"`
	StabilityPercent   float64 `json:"stabilityPercent"`
	TimeoutFrequency   float64 `json:"timeoutFrequency"`
}

// ScoreResult details overall endpoint evaluation and middlebox classification.
type ScoreResult struct {
	Numeric        int      `json:"numeric"`
	Grade          string   `json:"grade"`
	Classification string   `json:"classification"`
	Confidence     float64  `json:"confidence"`
	FalsePositive  bool     `json:"falsePositive"`
	Reasons        []string `json:"reasons"`
}

// ComputeProbeMetrics calculates statistical metrics across probe attempts.
func ComputeProbeMetrics(tcp TCPProbeResult, udp UDPProbeResult, dns DNSProbeResult) ProbeMetrics {
	var attempts []AttemptMetric
	attempts = append(attempts, tcp.Attempts...)
	attempts = append(attempts, udp.Attempts...)
	attempts = append(attempts, dns.Attempts...)

	var successful []int64
	timeouts := 0
	for _, a := range attempts {
		if a.Success {
			successful = append(successful, a.DurationMs)
		}
		if strings.EqualFold(a.ErrorCategory, "timeout") {
			timeouts++
		}
	}

	jitter := int64(0)
	if len(successful) > 1 {
		var sum float64
		for _, v := range successful {
			sum += float64(v)
		}
		mean := sum / float64(len(successful))
		var variance float64
		for _, v := range successful {
			delta := float64(v) - mean
			variance += delta * delta
		}
		jitter = int64(math.Round(math.Sqrt(variance / float64(len(successful)))))
	}

	rtt := tcp.MedianRTTMs
	if rtt == 0 {
		rtt = medianLatency(successful)
	}

	loss := 1.0 - calculateSuccessRatio(attempts)
	timeoutFreq := 0.0
	if len(attempts) > 0 {
		timeoutFreq = float64(timeouts) / float64(len(attempts))
	}

	stability := 100.0 * (1.0 - loss)
	if jitter > 250 {
		stability -= 15
	}
	if timeoutFreq > 0.25 {
		stability -= 20
	}
	if stability < 0 {
		stability = 0
	}

	return ProbeMetrics{
		RTTMs:              rtt,
		JitterMs:           jitter,
		PacketLossEstimate: roundTwoDec(loss * 100),
		StabilityPercent:   roundTwoDec(stability),
		TimeoutFrequency:   roundTwoDec(timeoutFreq * 100),
	}
}

// ScoreProbeResult scores tunnel viability and flags middlebox spoofing.
func ScoreProbeResult(
	tcp TCPProbeResult,
	tls TLSProbeResult,
	http HTTPProbeResult,
	ws WSProbeResult,
	quic QUICProbeResult,
	dns DNSProbeResult,
	metrics ProbeMetrics,
) ScoreResult {
	points := 0
	var reasons []string

	if tcp.Success {
		points += 25
		reasons = append(reasons, "tcp_connectivity")
	}
	if tcp.Consistency >= 0.67 {
		points += 15
		reasons = append(reasons, "retry_consistency")
	}
	if tls.Success {
		points += 20
		reasons = append(reasons, "tls_handshake")
	}
	if http.Success {
		points += 8
		reasons = append(reasons, "http_behavior")
	}
	if ws.Success {
		points += 10
		reasons = append(reasons, "websocket_upgrade")
	}
	if quic.Success {
		points += 8
		reasons = append(reasons, "quic_handshake")
	}
	if dns.UDPResponsive || dns.TCPResponsive {
		points += 6
		reasons = append(reasons, "dns_responsive")
	}

	if metrics.RTTMs > 0 && metrics.RTTMs < 150 {
		points += 5
	}
	if metrics.JitterMs < 80 {
		points += 5
	}
	if metrics.StabilityPercent >= 90 {
		points += 8
	}
	if points > 100 {
		points = 100
	}

	// False positive detection:
	// Middlebox responds to TCP SYN with ACK, but drops/blocks all higher protocols,
	// or exhibits erratic consistency (< 0.67).
	falsePositive := tcp.Success && tcp.Consistency < 0.67
	if tcp.Success && !tls.Success && !http.Success && !ws.Success && !quic.Success && !dns.UDPResponsive && !dns.TCPResponsive {
		falsePositive = true
	}

	classification := "Blocked"
	switch {
	case falsePositive:
		classification = "False Positive"
	case points >= 82 && tcp.Consistency >= 0.67 && (tls.Success || ws.Success || quic.Success):
		classification = "Tunnel Ready"
	case points >= 58:
		classification = "Partially Usable"
	case points >= 38:
		classification = "Unstable"
	}

	return ScoreResult{
		Numeric:        points,
		Grade:          calculateGrade(points),
		Classification: classification,
		Confidence:     roundTwoDec(calculateConfidence(tcp, tls, http, ws, quic, metrics, falsePositive)),
		FalsePositive:  falsePositive,
		Reasons:        reasons,
	}
}

// ParseCloudflareTrace parses egress trace responses (ip, loc).
func ParseCloudflareTrace(raw string) (string, string) {
	var ip string
	var countryCode string
	for _, line := range strings.Split(raw, "\n") {
		k, v, ok := strings.Cut(strings.TrimSpace(line), "=")
		if !ok {
			continue
		}
		switch strings.TrimSpace(k) {
		case "ip":
			ip = strings.TrimSpace(v)
		case "loc":
			countryCode = strings.ToUpper(strings.TrimSpace(v))
		}
	}
	return ip, countryCode
}

func calculateGrade(points int) string {
	switch {
	case points >= 94:
		return "A+"
	case points >= 85:
		return "A"
	case points >= 72:
		return "B"
	case points >= 58:
		return "C"
	case points >= 38:
		return "D"
	default:
		return "F"
	}
}

func calculateConfidence(
	tcp TCPProbeResult,
	tls TLSProbeResult,
	http HTTPProbeResult,
	ws WSProbeResult,
	quic QUICProbeResult,
	metrics ProbeMetrics,
	falsePositive bool,
) float64 {
	c := tcp.Consistency * 45
	if tls.Success {
		c += 20
	}
	if http.Success || ws.Success || quic.Success {
		c += 20
	}
	if metrics.StabilityPercent >= 85 {
		c += 15
	}
	if falsePositive {
		c -= 25
	}
	if c < 0 {
		return 0
	}
	if c > 100 {
		return 100
	}
	return c
}

func medianLatency(values []int64) int64 {
	if len(values) == 0 {
		return 0
	}
	cp := make([]int64, len(values))
	copy(cp, values)
	sort.Slice(cp, func(i, j int) bool { return cp[i] < cp[j] })
	return cp[len(cp)/2]
}

func calculateSuccessRatio(attempts []AttemptMetric) float64 {
	if len(attempts) == 0 {
		return 0
	}
	ok := 0
	for _, a := range attempts {
		if a.Success {
			ok++
		}
	}
	return float64(ok) / float64(len(attempts))
}

func roundTwoDec(v float64) float64 {
	return math.Round(v*100) / 100
}

File: src/packages/lumicore/src/scanner/tunnel_probe_evaluator.rs
//! Tunnel Probe Evaluator and False-Positive Middlebox Spoofing Detector.
//!
//! Evaluates multi-protocol tunnel probe results (TCP, TLS, HTTP, WebSocket, QUIC, DNS)
//! to calculate connectivity metrics (RTT, jitter, packet loss, stability), identify
//! ISP middlebox TCP-ACK spoofing (false positives), assign reliability grades (A+ to F),
//! and parse egress trace diagnostics.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AttemptMetric {
    pub success: bool,
    pub duration_ms: i64,
    pub error_category: Option<String>,
}

impl AttemptMetric {
    pub fn ok(duration_ms: i64) -> Self {
        Self {
            success: true,
            duration_ms,
            error_category: None,
        }
    }

    pub fn fail(duration_ms: i64, err: impl Into<String>) -> Self {
        Self {
            success: false,
            duration_ms,
            error_category: Some(err.into()),
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct TcpProbeResult {
    pub success: bool,
    pub attempts: Vec<AttemptMetric>,
    pub median_rtt_ms: i64,
    pub consistency: f64,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct TlsProbeResult {
    pub success: bool,
    pub handshake_ms: i64,
    pub version: Option<String>,
    pub cipher_suite: Option<String>,
    pub alpn: Option<String>,
    pub verified: bool,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HttpProbe {
    pub method: String,
    pub url: String,
    pub status_code: i32,
    pub duration_ms: i64,
    pub redirected: bool,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct HttpProbeResult {
    pub success: bool,
    pub probes: Vec<HttpProbe>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct WsProbeResult {
    pub success: bool,
    pub status_code: i32,
    pub duration_ms: i64,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct UdpProbeResult {
    pub reachable: bool,
    pub attempts: Vec<AttemptMetric>,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct QuicProbeResult {
    pub success: bool,
    pub handshake_ms: i64,
    pub alpn: Option<String>,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct DnsProbeResult {
    pub udp_responsive: bool,
    pub tcp_responsive: bool,
    pub answers: Vec<String>,
    pub attempts: Vec<AttemptMetric>,
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct ProbeMetrics {
    pub rtt_ms: i64,
    pub jitter_ms: i64,
    pub packet_loss_estimate: f64,
    pub stability_percent: f64,
    pub timeout_frequency: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ScoreResult {
    pub numeric: i32,
    pub grade: String,
    pub classification: String,
    pub confidence: f64,
    pub false_positive: bool,
    pub reasons: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct TunnelScanReport {
    pub endpoint: String,
    pub host: String,
    pub port: u16,
    pub tcp: TcpProbeResult,
    pub tls: TlsProbeResult,
    pub http: HttpProbeResult,
    pub websocket: WsProbeResult,
    pub udp: UdpProbeResult,
    pub quic: QuicProbeResult,
    pub dns: DnsProbeResult,
    pub metrics: ProbeMetrics,
    pub score: Option<ScoreResult>,
}

pub struct TunnelProbeEvaluator;

impl TunnelProbeEvaluator {
    /// Computes network connection metrics (RTT, jitter, loss, stability, timeouts)
    /// across collected TCP, UDP, and DNS probe attempts.
    pub fn compute_metrics(
        tcp: &TcpProbeResult,
        udp: &UdpProbeResult,
        dns: &DnsProbeResult,
    ) -> ProbeMetrics {
        let mut attempts: Vec<&AttemptMetric> = Vec::new();
        attempts.extend(tcp.attempts.iter());
        attempts.extend(udp.attempts.iter());
        attempts.extend(dns.attempts.iter());

        let mut successful_durations: Vec<i64> = Vec::new();
        let mut timeouts: usize = 0;

        for a in &attempts {
            if a.success {
                successful_durations.push(a.duration_ms);
            }
            if let Some(ref err) = a.error_category {
                if err.eq_ignore_ascii_case("timeout") {
                    timeouts += 1;
                }
            }
        }

        // Calculate jitter as sample standard deviation of successful latencies
        let jitter = if successful_durations.len() > 1 {
            let sum: f64 = successful_durations.iter().map(|&v| v as f64).sum();
            let mean = sum / (successful_durations.len() as f64);
            let variance: f64 = successful_durations
                .iter()
                .map(|&v| {
                    let d = (v as f64) - mean;
                    d * d
                })
                .sum::<f64>()
                / (successful_durations.len() as f64);
            variance.sqrt().round() as i64
        } else {
            0
        };

        // Determine RTT: prefer TCP median RTT, else median of successful probe latencies
        let rtt = if tcp.median_rtt_ms > 0 {
            tcp.median_rtt_ms
        } else {
            Self::median_latency(&successful_durations)
        };

        let total_attempts = attempts.len();
        let success_ratio = if total_attempts > 0 {
            let ok_count = attempts.iter().filter(|a| a.success).count();
            (ok_count as f64) / (total_attempts as f64)
        } else {
            0.0
        };

        let loss = 1.0 - success_ratio;
        let timeout_freq = if total_attempts > 0 {
            (timeouts as f64) / (total_attempts as f64)
        } else {
            0.0
        };

        let mut stability = 100.0 * (1.0 - loss);
        if jitter > 250 {
            stability -= 15.0;
        }
        if timeout_freq > 0.25 {
            stability -= 20.0;
        }
        if stability < 0.0 {
            stability = 0.0;
        }

        ProbeMetrics {
            rtt_ms: rtt,
            jitter_ms: jitter,
            packet_loss_estimate: Self::round_two_dec(loss * 100.0),
            stability_percent: Self::round_two_dec(stability),
            timeout_frequency: Self::round_two_dec(timeout_freq * 100.0),
        }
    }

    /// Evaluates protocol test outcomes to score endpoint quality, detect middlebox spoofing,
    /// calculate confidence, and assign classification grade.
    pub fn score_probe(
        tcp: &TcpProbeResult,
        tls: &TlsProbeResult,
        http: &HttpProbeResult,
        ws: &WsProbeResult,
        quic: &QuicProbeResult,
        dns: &DnsProbeResult,
        metrics: &ProbeMetrics,
    ) -> ScoreResult {
        let mut points = 0;
        let mut reasons = Vec::new();

        if tcp.success {
            points += 25;
            reasons.push("tcp_connectivity".to_string());
        }
        if tcp.consistency >= 0.67 {
            points += 15;
            reasons.push("retry_consistency".to_string());
        }
        if tls.success {
            points += 20;
            reasons.push("tls_handshake".to_string());
        }
        if http.success {
            points += 8;
            reasons.push("http_behavior".to_string());
        }
        if ws.success {
            points += 10;
            reasons.push("websocket_upgrade".to_string());
        }
        if quic.success {
            points += 8;
            reasons.push("quic_handshake".to_string());
        }
        if dns.udp_responsive || dns.tcp_responsive {
            points += 6;
            reasons.push("dns_responsive".to_string());
        }

        if metrics.rtt_ms > 0 && metrics.rtt_ms < 150 {
            points += 5;
        }
        if metrics.jitter_ms < 80 {
            points += 5;
        }
        if metrics.stability_percent >= 90.0 {
            points += 8;
        }

        if points > 100 {
            points = 100;
        }

        // False positive detection:
        // Case 1: TCP succeeds, but retry consistency is low (< 0.67) indicating erratic ACK injection.
        // Case 2: TCP succeeds, but TLS, HTTP, WebSocket, QUIC, and DNS ALL fail. This indicates
        // middlebox SYN-ACK spoofing where TCP connects but higher application layers are blocked/dropped.
        let mut false_positive = tcp.success && tcp.consistency < 0.67;
        if tcp.success
            && !tls.success
            && !http.success
            && !ws.success
            && !quic.success
            && !dns.udp_responsive
            && !dns.tcp_responsive
        {
            false_positive = true;
        }

        let classification = if false_positive {
            "False Positive".to_string()
        } else if points >= 82 && tcp.consistency >= 0.67 && (tls.success || ws.success || quic.success) {
            "Tunnel Ready".to_string()
        } else if points >= 58 {
            "Partially Usable".to_string()
        } else if points >= 38 {
            "Unstable".to_string()
        } else {
            "Blocked".to_string()
        };

        let grade = Self::calculate_grade(points);
        let conf = Self::calculate_confidence(tcp, tls, http, ws, quic, metrics, false_positive);

        ScoreResult {
            numeric: points,
            grade,
            classification,
            confidence: Self::round_two_dec(conf),
            false_positive,
            reasons,
        }
    }

    /// Evaluates an entire scan report and updates its metrics and score.
    pub fn evaluate_report(report: &mut TunnelScanReport) {
        let metrics = Self::compute_metrics(&report.tcp, &report.udp, &report.dns);
        report.metrics = metrics.clone();
        let score = Self::score_probe(
            &report.tcp,
            &report.tls,
            &report.http,
            &report.websocket,
            &report.quic,
            &report.dns,
            &metrics,
        );
        report.score = Some(score);
    }

    /// Parses Cloudflare egress trace data (`https://www.cloudflare.com/cdn-cgi/trace`)
    /// extracting `ip` and uppercase `loc` (ISO alpha-2 country code).
    pub fn parse_cloudflare_trace(raw: &str) -> (Option<String>, Option<String>) {
        let mut ip = None;
        let mut loc = None;

        for line in raw.lines() {
            let line = line.trim();
            if let Some((k, v)) = line.split_once('=') {
                let key = k.trim();
                let val = v.trim();
                match key {
                    "ip" => {
                        if !val.is_empty() {
                            ip = Some(val.to_string());
                        }
                    }
                    "loc" => {
                        if !val.is_empty() {
                            loc = Some(val.to_ascii_uppercase());
                        }
                    }
                    _ => {}
                }
            }
        }
        (ip, loc)
    }

    fn calculate_grade(points: i32) -> String {
        match points {
            p if p >= 94 => "A+".to_string(),
            p if p >= 85 => "A".to_string(),
            p if p >= 72 => "B".to_string(),
            p if p >= 58 => "C".to_string(),
            p if p >= 38 => "D".to_string(),
            _ => "F".to_string(),
        }
    }

    fn calculate_confidence(
        tcp: &TcpProbeResult,
        tls: &TlsProbeResult,
        http: &HttpProbeResult,
        ws: &WsProbeResult,
        quic: &QuicProbeResult,
        metrics: &ProbeMetrics,
        false_positive: bool,
    ) -> f64 {
        let mut c = tcp.consistency * 45.0;
        if tls.success {
            c += 20.0;
        }
        if http.success || ws.success || quic.success {
            c += 20.0;
        }
        if metrics.stability_percent >= 85.0 {
            c += 15.0;
        }
        if false_positive {
            c -= 25.0;
        }
        if c < 0.0 {
            c = 0.0;
        }
        if c > 100.0 {
            c = 100.0;
        }
        c
    }

    fn median_latency(values: &[i64]) -> i64 {
        if values.is_empty() {
            return 0;
        }
        let mut sorted = values.to_vec();
        sorted.sort_unstable();
        sorted[sorted.len() / 2]
    }

    fn round_two_dec(val: f64) -> f64 {
        (val * 100.0).round() / 100.0
    }
}
